src/algorithms/semihdp_sampler.hpp

```diff
@@ -0,0 +1,205 @@
+ * semi-hierarchical Dirichlet process in [1].
+ * 
+ * Extra goodies: we can now tune the pseudo-prior generation, by randomly
+ * drawing the cardinalities from a multinomial distribution, and perturbing
+ * the atoms of the mixing measure.
+ *
+ * [1]: "The semi-hierarchical Dirichlet Process and its application to
+ * clustering homogeneous distributions", Beraha, Guglielmi and Quintana
+ * arXiv: 2005.10287
+ */
+
+class SemiHdpSampler {
+ protected:
+  bayesmix::SemiHdpParams params;
+
+  std::vector<MatrixXd> data;  // one vector per group
+  int ngroups;
+  std::vector<int> n_by_group;
+
+  std::shared_ptr<BaseHierarchy> G0_master_hierarchy;
+  std::shared_ptr<BaseHierarchy> G00_master_hierarchy;
+
+  // these should be just copies
+  std::vector<std::vector<std::shared_ptr<BaseHierarchy>>> theta_star;
+  // idiosincratic tables
+  std::vector<std::vector<std::shared_ptr<BaseHierarchy>>> theta_tilde;
+  // shared tables
+  std::vector<std::shared_ptr<BaseHierarchy>> taus;
+  // counts
+  std::vector<std::vector<int>> n_by_theta_star;
+
+  // stuff for pseudoprior
+  std::vector<std::vector<std::shared_ptr<BaseHierarchy>>> theta_star_pseudo;
+  std::vector<std::vector<int>> n_by_theta_star_pseudo;
+
+  // theta_{ij} = theta_{c_i, s_{ij}}
+  std::vector<std::vector<int>> s;
+  // which restaurant each grups enters from
+  std::vector<int> c;
+  std::vector<bool> is_used_c;
+  VectorXd omega;
+  std::vector<std::vector<int>> t, v;
+
+  // number of theta_stars equal to tau_h across all groups
+  std::vector<int> m;
+
+  double w = 0.5;
+  double alpha, gamma, a_w, b_w;
+
+  std::vector<MemoryCollector<MarginalState>> pseudoprior_collectors;
+  int pseudo_iter;
+
+  bool adapt = false;
+  std::string c_update;
+
+ public:
+  SemiHdpSampler() {}
+  ~SemiHdpSampler() {}
+
+  SemiHdpSampler(const std::vector<MatrixXd> &data,
+                 std::shared_ptr<BaseHierarchy> hier,
+                 bayesmix::SemiHdpParams params);
+
+  void initialize();
+
+  void step() {
+    update_unique_vals();
+    update_w();
+    if (!adapt) {
+      sample_pseudo_prior();
+      update_c();
+    }
+    update_omega();
+    update_t();
+    update_s();
+    relabel();
+  }
+
+  void run(int adapt_iter, int burnin, int iter, int thin,
+           BaseCollector<bayesmix::SemiHdpState> *collector,
+           const std::vector<MemoryCollector<MarginalState>>
+               &pseudoprior_collectors) {
+    this->pseudoprior_collectors = pseudoprior_collectors;
+    std::cout << "Run, number of pseudoprior_collectors: "
+              << this->pseudoprior_collectors.size() << std::endl;
+    pseudo_iter = pseudoprior_collectors[0].get_size();
+
+    initialize();
+    update_unique_vals();
+
+    if (adapt_iter > 0) {
+      adapt = true;
+      for (int i = 0; i < adapt_iter; i++) {
+        step();
+        if ((i + 1) % 100 == 0) {
+          std::cout << "Adapt iter: " << i << " / " << adapt_iter << std::endl;
```

Comment:
A DEBUG string is needed here and below with some other prints

Reply:
I solved this with a "displayprogress" flag. I want to keep the progress displayed since it might run for several hours (on my laptop) and even when I'm not debugging I want to know how long it's going to take